src/scalaDebugger.ts

```diff
@@ -46,15 +50,169 @@ export async function start(
+        if (className === undefined) return undefined;
+        return this.askForConfigurationName(className).then((name) => {
+          if (name === undefined) return undefined;
+          const result: DebugConfiguration = {
+            type: configurationType,
+            name: name,
+            request: launchRequestType,
+            testClass: className,
+            buildTarget: buildTarget,
+          };
+          return result;
+        });
+      });
+    });
+  }
+
+  private askForBuildTarget(): Thenable<string | null | undefined> {
+    return vscode.window
+      .showInputBox({
+        prompt: "Enter name of the build target",
+        placeHolder: "Optional, you can leave it empty",
+      })
+      .then((buildTarget) => {
+        if (buildTarget === undefined) {
+          return undefined;
+        } else if (buildTarget === "") {
+          return null;
+        } else {
+          return buildTarget;
+        }
+      });
+  }
+
+  private askForClassName(): Thenable<string | undefined> {
+    return vscode.window.showInputBox({
+      prompt: "Enter name of the class to debug",
+      placeHolder: "<package>.<Class>",
+    });
+  }
+
+  private askForConfigurationName(
+    className: string
+  ): Thenable<string | undefined> {
+    return vscode.window.showInputBox({
+      prompt: "Enter name of the configuration",
+      value: `Debug ${className}`,
+    });
+  }
+
+  private askForArguments(): Thenable<string[] | undefined> {
+    return vscode.window
+      .showInputBox({ prompt: "Enter argument or leave it empty" })
+      .then((argument) => {
+        if (argument === undefined) {
+          return undefined;
+        } else if (argument === "") {
+          return [];
+        } else {
+          return this.askForArguments().then((rest) => {
+            if (rest === undefined) {
+              return [argument];
+            } else {
+              rest.unshift(argument);
+              return rest;
+            }
+          });
+        }
+      });
+  }
+}
+
+class ScalaDebugServerFactory implements vscode.DebugAdapterDescriptorFactory {
+  createDebugAdapterDescriptor(
+    session: vscode.DebugSession
+  ): ProviderResult<DebugAdapterDescriptor> {
+    if (
+      session.configuration.mainClass !== undefined ||
+      session.configuration.testClass !== undefined
+    ) {
+      return vscode.commands
+        .executeCommand<DebugSession>(
+          startAdapterCommand,
+          session.configuration
+        )
+        .then((debugSession) => {
+          if (debugSession === undefined) return null;
+          return debugServerFromUri(debugSession.uri);
+        });
+    }
+    return null;
+  }
+}
+
+export function debugServerFromUri(uri: string): vscode.DebugAdapterServer {
+  const debugServer = vscode.Uri.parse(uri);
+  const segments = debugServer.authority.split(":");
+  return new vscode.DebugAdapterServer(
+    parseInt(segments[segments.length - 1]),
+    segments[0]
```

Comment:
can you give meaningful names to the segments?

Something like:

```suggestion
  const segments = debugServer.authority.split(":");
  const protocol = segments[0]
  const port = parseInt(segments[segments.length - 1]);
  return new vscode.DebugAdapterServer(port, protocol);
```

(provided that `port` and `protocol` make sense as names)

Reply:
done. I've put `host` instead of `protocol` -- at least it corresponds with `DebugAdapterServer` parameters.